### PR TITLE
Certparty

### LINKED ISSUE #140: Add names field
collect all names that the certificate is valid for and put them in one place

### PR DESCRIPTION
**Pull request by shoegaze**
Closed #140.

### CODE REVIEW

File: ztools/x509/json.go [hunk 1]
```diff
@@ -202,8 +201,10 @@ type jsonCertificate struct {
 	FingerprintMD5            CertificateFingerprint       `json:"fingerprint_md5"`
 	FingerprintSHA1           CertificateFingerprint       `json:"fingerprint_sha1"`
 	FingerprintSHA256         CertificateFingerprint       `json:"fingerprint_sha256"`
+	SPKIFingerprint           CertificateFingerprint       `json:"spki_fingerprint"`
```
Comment (dadrian) [inline ztools/x509/json.go]: Wut.

File: ztools/x509/json.go [hunk 2]
```diff
@@ -167,11 +167,10 @@ func (v *validity) UnmarshalJSON(b []byte) error {
...
 type jsonSubjectKeyInfo struct {
-	KeyAlgorithm    PublicKeyAlgorithm     `json:"key_algorithm"`
-	RSAPublicKey    *keys.RSAPublicKey     `json:"rsa_public_key,omitempty"`
-	DSAPublicKey    interface{}            `json:"dsa_public_key,omitempty"`
-	ECDSAPublicKey  interface{}            `json:"ecdsa_public_key,omitempty"`
-	SPKIFingerprint CertificateFingerprint `json:"fingerprint_sha256"`
+	KeyAlgorithm   PublicKeyAlgorithm `json:"key_algorithm"`
+	RSAPublicKey   *keys.RSAPublicKey `json:"rsa_public_key,omitempty"`
+	DSAPublicKey   interface{}        `json:"dsa_public_key,omitempty"`
+	ECDSAPublicKey interface{}        `json:"ecdsa_public_key,omitempty"`
```
Comment (dadrian) [inline ztools/x509/json.go]: Wut.

File: ztools/x509/json.go [hunk 3]
```diff
@@ -220,7 +221,16 @@ func (c *Certificate) MarshalJSON() ([]byte, error) {
 	jc.Subject = c.Subject
 	jc.SubjectDN = c.Subject.String()
 	jc.SubjectKeyInfo.KeyAlgorithm = c.PublicKeyAlgorithm
-	jc.SubjectKeyInfo.SPKIFingerprint = c.SPKIFingerprint
+
+	// 1 for the DN, and then however many DNS names there are
+	jc.Names = make([]string, 1+len(c.DNSNames))
+
+	jc.Names[0] = jc.SubjectDN
```
Comment (dadrian) [inline ztools/x509/json.go]: The `SubjectDN` isn't a URL. I think you'll need to go through each of `[]Names` in `pkix.Name`, check to see if it's a URL, and add it if it is. Golang's `net.URL` probably helps with this.

File: ztools/x509/json.go (hunk 2 above)
Comment (shoegaze) [inline ztools/x509/json.go]: You see, I asked @zakird to check my sanity...

File: ztools/x509/json.go (hunk 3 above)
Comment (shoegaze) [inline ztools/x509/json.go]: Ah. My brain interpreted "DN" as domain name, not distinguished name.

Comment (shoegaze) [inline ztools/x509/json.go]: Would it make more sense to reconstruct them from the domain component?

File: ztools/x509/json.go [hunk 4]
```diff
@@ -221,7 +223,25 @@ func (c *Certificate) MarshalJSON() ([]byte, error) {
...
+			}
+		}
+	}
+
+	for _, name := range c.DNSNames {
+
+		if _, err := url.Parse(name); err == nil {
+			jc.Names = append(jc.Names, name)
+		}
+	}
```
Comment (dadrian) [inline ztools/x509/json.go]: What happens when one of these contains `*.example.com`?

Comment (shoegaze) [inline ztools/x509/json.go]: https://play.golang.org/p/kV7FopPrh0

Comment (shoegaze) [inline ztools/x509/json.go]: Actually, as it turns out net/url is the wrong library to validate urls. Switching to govalidate, see here. https://www.socketloop.com/tutorials/golang-how-to-validate-url-the-right-way

Comment (shoegaze) [inline ztools/x509/json.go]: Grr. Except this doesn't work either, apparently.

**Comment by shoegaze** [comment]
bumping this, I think it's ready to go.

File: ztools/x509/json.go [hunk 5]
```diff
@@ -221,7 +224,33 @@ func (c *Certificate) MarshalJSON() ([]byte, error) {
...
+				jc.Names = append(jc.Names, name)
+			}
+		}
+	}
+
+	for _, name := range c.DNSNames {
+
+		if name[0] == '*' {
+			name = name[2:]
+		}
```
Comment (zakird) [inline ztools/x509/json.go]: Do just lose the wildcard here then? Shouldn't something like `*.zmap.io` be what's added?

Comment (shoegaze) [inline ztools/x509/json.go]: ah, good catch. Fixing.

**Comment by zakird** [comment]
We should also add this to the zgrab and ztag schemas.

**Comment by shoegaze** [comment]
I'm working on adding more test data. I'll get back to you.

**Comment by shoegaze** [comment]
Okay, now I have tested more thoroughly.

**Comment by zakird** [comment]
Sounds good. I'm with merging once the field is in the schema.

File: ztools/x509/json.go [hunk 6]
```diff
@@ -33,31 +35,31 @@ type auxKeyUsage struct {
 func (k KeyUsage) MarshalJSON() ([]byte, error) {
 	var enc auxKeyUsage
 	enc.Value = uint32(k)
-	if k & KeyUsageDigitalSignature > 0 {
+	if k&KeyUsageDigitalSignature > 0 {
```
Comment (zzma) [inline ztools/x509/json.go]: Nitpick: we should not undo the proper formatting here

Comment (shoegaze) [inline ztools/x509/json.go]: This was done by go fmt

Comment (zzma) [inline ztools/x509/json.go]: Oh true. That's weird, looks like my editor auto-formats things differently from `go fmt`. I'll fix my setup

**Pull request merged**